### PR TITLE
feat: add API authentication support

### PR DESCRIPTION
**Pull request by sgrove**
A continuation of the work done in #104 that adds in dynamic fetching of secrets, and lays the groundwork for some backend changes coming to how the secrets will be passed in to functions in the future.

### CODE REVIEW

File: .eslintrc.js
Comment (ascorbic) [inline .eslintrc.js:7]: The changes to this file can be removed

File: src/function/handler.ts
Comment (ascorbic) [inline src/function/handler.ts:10]:
```suggestion
export interface Handler<C extends Context = Context> {
```

This needs a default

File: src/lib/secrets.js
Comment (ascorbic) [inline src/lib/secrets.js:1]: The rest of the package is now written in TypeScript, so it would make sense to do the same here. As you already have the typings in a .d.ts it hopefully wouldn't be too much of a lift.

File: src/lib/secrets.js
Comment (ascorbic) [inline src/lib/secrets.js:70]: You can use optional chaining here

File: src/lib/secrets.js
Comment (ascorbic) [inline src/lib/secrets.js:91]: Optional chaining here

**Review by ascorbic — COMMENTED** [review]
Looking good! I haven't tested the typings though, so they will need to be looked at.

File: src/lib/secrets.js
Comment (sgrove) [reply src/lib/secrets.js:1]: Converting this to TypeScript is proving to be fairly painful. Is there some internal guide for e.g. how to get around the `ts-ignore` ban and whatnot?

Comment (ascorbic) [reply src/lib/secrets.js:1]: OK, in the interest of getting this out of the door on a deadline I suggest shipping it as js today, and we can miograte it when we have a bit more time

Comment (sgrove) [reply src/lib/secrets.js]: Whoops, missed this, and went through quite a large rewrite - could use another pair of eyes to look it over!

**Review by ascorbic — APPROVED** [review]
LGTM, but I'd like a serverless pod review before merging

File: src/lib/secrets_helper.ts
Comment (eduardoboucas) [inline src/lib/secrets_helper.ts:124]: I think this would be a bit easier to read if we reversed the condition of the if statement and did an early return, but this boils down to personal preference, so by all means feel free to ignore.

**Review by eduardoboucas — DISMISSED** [review]
LGTM! I left a few comments, but none of them are blocking.

File: src/lib/secrets.ts
Comment (ascorbic) [reply src/lib/secrets.ts:21]: I think the idea here is that it _does_ check if the prop exists, and falls-back to using the env var if it's not available.

Comment (eduardoboucas) [reply src/lib/secrets.ts:21]: Ah, of course! I guess I got confused by the `await`. But in that case, do we need a check at all? If we always pass `event` to `getSecrets` and the event doesn't have the `_oneGraph` property, we'll be able to handle that in https://github.com/netlify/functions/pull/217/files#diff-c8a3442b81d0b5521bfd6b9591f5ede62cde7fdd943b525e463533b3c77721adR147?

Comment (ascorbic) [reply src/lib/secrets.ts:21]: That's a good point. Just pass it through, and the check will catch it.

File: src/lib/secrets_helper.ts
Comment (sgrove) [reply src/lib/secrets_helper.ts:55]: For the time being, I'll keep the 4 LoC for this function inline. Happy to revisit it after today though!

File: src/lib/secrets.ts
Comment (sgrove) [reply src/lib/secrets.ts:21]: Sounds good! I was struggling with TypeScript, and I think this is what made it happy - but I was one step away from making it a bit simpler as you suggested!

**Review by eduardoboucas — APPROVED** [review]
🚀

**Review by ascorbic — APPROVED** [review]
:shipit:

**Pull request merged**
